### DISCUSSION TITLE
Conditional `with_loader_criteria` query option

### DISCUSSION THREAD

**Discussion by decaz** [sqlalchemy/sqlalchemy#6852] [Usage Questions] [answered]
I use `with_loader_criteria` query option to implement "soft delete" pattern ([docs with example](https://docs.sqlalchemy.org/en/14/orm/session_events.html#do-orm-execute-global-criteria)) and there is need to not just pass `include_deleted=True` execution option but to pass certain mapped classes whose deleted objects should be included.

I have `SoftDelete` mixin class with `is_deleted` column and apply option within `do_orm_execute` hook in the following way:
```python
include_deleted = execute_state.execution_options.get('include_deleted', False)
if not include_deleted:
    execute_state.statement = execute_state.statement.options(
        with_loader_criteria(SoftDelete, lambda cls: ~cls.is_deleted, include_aliases=True)
    )
```

How can I use `with_loader_criteria` query option to call something like this? (`A`, `B` and `C` are all inerited from `SoftDelete`)
```python
all_A_B_C_except_deleted_B = session.query(A, B, C).execution_options(include_deleted=(A, C)).all()
```

**Comment by zzzeek** [comment, marked as answer]
hi -

why dont you specify with_loader_criteria just for B?   I dont think im understanding the problem.

For example if you had `execution_options(include_deleted=(A, C))`:

```py
include_deleted = execute_state.execution_options.get('include_deleted', False)
if not include_deleted:
    execute_state.statement = execute_state.statement.options(
        with_loader_criteria(SoftDelete, lambda cls: ~cls.is_deleted, include_aliases=True)
    )
else:
    execute_state.statement = execute_state.statement.options(*[
          with_loader_criteria(sub, ~sub.is_deleted, include_aliases=True)
          for sub in SoftDelete.__subclasses__() if sub not in include_deleted
    ]
    )

```

**Reply by decaz** [reply]
@zzzeek thank you! I didn't event think about `__subclasses__` method..
